src/ai/defaultai.cc

```diff
@@ -3350,6 +3369,61 @@ void DefaultAI::diplomacy_actions(const Time& gametime) {
                                           (accept ? Widelands::DiplomacyAction::kAcceptJoin :
                                                     Widelands::DiplomacyAction::kRefuseJoin)),
 			                             pda.sender);
+			verb_log_dbg_time(
+			   gametime,
+			   "AI Diplomacy: Player(%d), %s the invitation of player (%d) with diploscore: %d\n",
+			   static_cast<unsigned int>(pda.other), accept ? "accepts" : "denies",
+			   static_cast<unsigned int>(pda.sender), player_statistics.get_diplo_score(pda.sender));
+			if (accept) {
+				request_accepted = true;
+			}
+		}
+	}
+	if (request_accepted || me_def) {
+		return;
+	}
+
+	for (Widelands::PlayerNumber opn = 1; opn <= game().map().get_nrplayers(); ++opn) {
+		const Widelands::Player* other_player = game().get_player(opn);
+		// other player needs to exist and different from us
+		// we need to be still alive and we don't have send a request in last 5 + X minutes
+		if (other_player == nullptr || opn == mypn ||
+		    player_statistics.player_diplo_requested_lately(opn, gametime)) {
+			continue;
+		}
+		if (player_statistics.get_diplo_score(opn) >= 40) {
+			if (((other_player->team_number() == 0 ||
+			      player_statistics.team_power(me->team_number()) >
+			         player_statistics.team_power(other_player->team_number())) &&
+			     player_statistics.members_in_team(me->team_number()) <
+			        player_statistics.players_active() - 1) ||
+			    player_statistics.members_in_team(other_player->team_number()) ==
+			       player_statistics.players_active() - 1) {
+				game().send_player_diplomacy(mypn, Widelands::DiplomacyAction::kInvite, opn);
+				player_statistics.set_last_time_requested(gametime, opn);
+				verb_log_dbg_time(
+				   gametime,
+				   "AI Diplomacy: Player(%d), invites player (%d) to join with diploscore: %d\n",
+				   static_cast<unsigned int>(mypn), static_cast<unsigned int>(opn),
+				   player_statistics.get_diplo_score(opn));
+			} else if (other_player->team_number() != me->team_number()) {
+				game().send_player_diplomacy(mypn, Widelands::DiplomacyAction::kJoin, opn);
+				player_statistics.set_last_time_requested(gametime, opn);
+				verb_log_dbg_time(
+				   gametime,
+				   "AI Diplomacy: Player(%d), requests to join player (%d) with diploscore: %d\n",
+				   static_cast<unsigned int>(mypn), static_cast<unsigned int>(opn),
+				   player_statistics.get_diplo_score(opn));
+			}
+		} else if (player_statistics.get_diplo_score(opn) < -15 &&
+		           other_player->team_number() == me->team_number() &&
+		           other_player->team_number() != 0) {
+			game().send_player_diplomacy(mypn, Widelands::DiplomacyAction::kLeaveTeam, opn);
```

Comment:
Could we loop here through all other team members and reset `last_time_requested` to allow inviting over into a new team the good ones quickly?

Reply:
that might result in a lot of requests to human players from computerplayers that just left the team.   
Remember that one goal is not to spam human players with requests.

Reply:
I don't think that would happen all that often. It's not very likely that you'd often end up in a team with both strong and weak players. But yes, this would pretty much require the diploscore based invite/join distinction to prevent flip-flopping.

It's actually a shortcoming of the current diplomacy possibilities that it's not possible to kick out a weak player from a strong team, so this workaround is needed to achieve it. (it would be a more complicated UI though – I almost created a feature request for such a thing yesterday, but then thought better of it)

Reply:
Well, this is indeed less critical. The other half of the solution is improving invite/join, so that players don't have to join teams with players they don't like in order to get allied with those who they like.